Use a separate bag per test

zmichaels11 opened the pull request:
There's a bug where E2E tests will begin with a bagfile already existing.
This changes the recording E2E tests to use a separate path per test which reduces test flakiness.

Signed-off-by: Zachary Michaels <zmichaels11@gmail.com>

zmichaels11 commented:
Gist: https://gist.githubusercontent.com/zmichaels11/d9c32806fc127d14b4abacaab82df97f/raw/7bef5e3cd5bbaa37cb7441644398accefbaa79ec/ros2.repos
BUILD args: --packages-up-to rosbag2_tests
TEST args: --packages-select rosbag2_tests
Job: ci_launcher

zmichaels11 commented:
Running CI on windows with `retest-until-fail 50`. Will update with badge once task begins.

zmichaels11 commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=9217)](http://ci.ros2.org/job/ci_linux/9217/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=4908)](http://ci.ros2.org/job/ci_linux-aarch64/4908/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=7518)](http://ci.ros2.org/job/ci_osx/7518/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=9136)](http://ci.ros2.org/job/ci_windows/9136/)
* Windows-container [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows-container&build=16)](http://ci.ros2.org/job/ci_windows-container/16/)

zmichaels11 commented on rosbag2_tests/test/rosbag2_tests/test_rosbag2_record_end_to_end.cpp:
```diff
@@ -200,7 +205,7 @@ TEST_F(RecordFixture, record_end_to_end_with_splitting_bagsize_split_is_at_least
     for (int i = 0; i < expected_splits; ++i) {
       std::stringstream bagfile_name;
-      bagfile_name << "bag_" << i << ".db3";
+      bagfile_name << bagfile_name << i << ".db3";
```
`bagfile_name` is both the `stringstream` and const c-string

zmichaels11 commented on rosbag2_tests/test/rosbag2_tests/test_rosbag2_record_end_to_end.cpp:
```diff
@@ -55,6 +55,7 @@ std::shared_ptr<test_msgs::msg::Strings> create_string_message(
 }  // namespace
 
 TEST_F(RecordFixture, record_end_to_end_test) {
+  set_path("record");
```
extract constant

zmichaels11 commented on rosbag2_tests/test/rosbag2_tests/test_rosbag2_record_end_to_end.cpp:
```diff
@@ -332,7 +343,7 @@ TEST_F(RecordFixture, record_end_to_end_with_splitting_splits_bagfile) {
-      bag_name << "bag_" << i << ".db3";
+      bag_name << bagfile_name << i << ".db3";
 
       const auto bag_path = rcpputils::fs::path(root_bag_path_) / bag_name.str();
```
`stringstream` is `filename` elsewhere

piraka9011 commented:
New windows build: 
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=9148)](http://ci.ros2.org/job/ci_windows/9148/)

zmichaels11 commented on rosbag2_tests/test/rosbag2_tests/test_rosbag2_record_end_to_end.cpp:
```diff
@@ -83,7 +85,8 @@ TEST_F(RecordFixture, record_end_to_end_test) {
   metadata.storage_identifier = "sqlite3";
-  metadata.relative_file_paths = {"bag_0.db3"};
+  std::string bag_file_path = std::string(bag_file_name) + "_0.db3";
+  metadata.relative_file_paths = {bag_file_name};
```
Should this be `bag_file_path`?

zmichaels11 commented on rosbag2_tests/test/rosbag2_tests/test_rosbag2_record_end_to_end.cpp:
```diff
@@ -83,7 +85,8 @@ TEST_F(RecordFixture, record_end_to_end_test) {
   metadata.version = 1;
   metadata.storage_identifier = "sqlite3";
-  metadata.relative_file_paths = {"bag_0.db3"};
+  std::string bag_file_path = std::string(bag_file_name) + "_0.db3";
```
`database_path_` points to the relative path to the first db.

zmichaels11 commented on rosbag2_tests/test/rosbag2_tests/test_rosbag2_record_end_to_end.cpp:
```diff
@@ -199,14 +206,11 @@ TEST_F(RecordFixture, record_end_to_end_with_splitting_bagsize_split_is_at_least
-        metadata.relative_file_paths.push_back(bagfile_path.string());
+      std::stringstream bag_name;
+      bag_name << bag_file_name << "_" << i << ".db3";
+      const auto bag_file_path = rcpputils::fs::path(root_bag_path_) / bag_name.str();
```
```suggestion
      const auto bag_file_path = rcpputils::fs::path{root_bag_path} / bag_name.str();
```

zmichaels11 commented on rosbag2_tests/test/rosbag2_tests/test_rosbag2_record_end_to_end.cpp:
```diff
@@ -227,6 +231,9 @@ TEST_F(RecordFixture, record_end_to_end_with_splitting_bagsize_split_is_at_least
   for (int i = 0; i < actual_splits - 1; ++i) {
     const auto bagfile_path = metadata.relative_file_paths[i];
+
+    std::cout << "Checking: " << bagfile_path << std::endl;
```
Replace with log statement

zmichaels11 commented on rosbag2_tests/test/rosbag2_tests/test_rosbag2_record_end_to_end.cpp:
```diff
@@ -273,7 +282,9 @@ TEST_F(RecordFixture, record_end_to_end_with_splitting_max_size_not_reached) {
-    const auto bag_path = rcpputils::fs::path(root_bag_path_) / "bag_0.db3";
+    std::stringstream file_name;
+    file_name << bag_file_name << "_0.db3";
+    const auto bag_path = rcpputils::fs::path(root_bag_path_) / file_name.str();
```
```suggestion
    const auto bag_path = rcpputils::fs::path{root_bag_path_} / file_name.str();
```

dabonnie commented on rosbag2_tests/test/rosbag2_tests/record_fixture.hpp:
```diff
@@ -45,18 +45,22 @@ class RecordFixture : public TemporaryDirectoryFixture
   }
 
-  void SetUp() override
+  void set_path(const std::string & bagname)
```
Any reason for the method syntax change? https://google.github.io/styleguide/cppguide.html#Function_Names

piraka9011 replied:
`SetUp()` is a method specific to [GTest](https://github.com/google/googletest/blob/master/googletest/docs/primer.md#test-fixtures-using-the-same-data-configuration-for-multiple-tests-same-data-multiple-tests) that is usually overridden to setup a fixture. `set_path()` is a method we introduced.

zmichaels11 commented:
Addressed all issues; rebuilding CI:
* Windows: [![Build Status](https://ci.ros2.org/job/ci_windows/9158/badge/icon)](https://ci.ros2.org/job/ci_windows/9158/)

dabonnie commented on rosbag2_tests/test/rosbag2_tests/record_fixture.hpp:
```diff
@@ -45,18 +45,22 @@ class RecordFixture : public TemporaryDirectoryFixture
+    auto root_path = rcpputils::fs::path{temporary_dir_path_} / bagname;
+
+    if (root_path.exists()) {
+      remove_directory_recursively(root_path.string());
```
Shouldn't the removal happen in TearDown or at the end of each test that called set_path? Seems strange to check if resources exist that weren't created by a brand new test (RAII).

zmichaels11 replied:
I'm addressing this in the source as well.

The test fixture already does define an RAII cleanup on TestFixture dtor. This is ensuring that the path specified is empty.

thomas-moulard replied:
+1 to Devin's comment, can't we do the clean up in `SetUp` ?

zmichaels11 replied:
I'm switching to `ASSERT_FALSE` on exist since cleanup should already be handled.

thomas-moulard commented on rosbag2_tests/test/rosbag2_tests/test_rosbag2_record_end_to_end.cpp:
```diff
@@ -55,6 +55,8 @@ std::shared_ptr<test_msgs::msg::Strings> create_string_message(
 
 TEST_F(RecordFixture, record_end_to_end_test) {
+  const auto bag_file_name = get_test_name();
+  set_path(bag_file_name);
```
Could we do this in `SetUp`?

zmichaels11 replied:
I actually don't know. I'll give it a  try and report back...

zmichaels11 commented:
Closing in favor of PR #286
